drivers/crypto/crypto_stm32.c

```diff
@@ -0,0 +1,400 @@
+
+	ctx_idx = crypto_stm32_get_unused_session_index();
+	if (ctx_idx < 0) {
+		LOG_ERR("No free session for now");
+		return -ENOSPC;
+	}
+	session = &crypto_stm32_sessions[ctx_idx];
+	memset(&session->config, 0, sizeof(session->config));
+
+	switch (ctx->keylen) {
+	case 16U:
+		session->config.KeySize = CRYP_KEYSIZE_128B;
+		break;
+	case 24U:
+		session->config.KeySize = CRYP_KEYSIZE_192B;
+		break;
+	case 32U:
+		session->config.KeySize = CRYP_KEYSIZE_256B;
+		break;
+	}
+
+	if (op_type == CRYPTO_CIPHER_OP_ENCRYPT) {
+		switch (mode) {
+		case CRYPTO_CIPHER_MODE_ECB:
+			session->config.Algorithm = CRYP_AES_ECB;
+			ctx->ops.block_crypt_hndlr = crypto_stm32_ecb_encrypt;
+			break;
+		case CRYPTO_CIPHER_MODE_CBC:
+			session->config.Algorithm = CRYP_AES_CBC;
+			ctx->ops.cbc_crypt_hndlr = crypto_stm32_cbc_encrypt;
+			break;
+		case CRYPTO_CIPHER_MODE_CTR:
+			session->config.Algorithm = CRYP_AES_CTR;
+			ctx->ops.ctr_crypt_hndlr = crypto_stm32_ctr_encrypt;
+			break;
+		default:
+			break;
+		}
+	} else {
+		switch (mode) {
+		case CRYPTO_CIPHER_MODE_ECB:
+			session->config.Algorithm = CRYP_AES_ECB;
+			ctx->ops.block_crypt_hndlr = crypto_stm32_ecb_decrypt;
+			break;
+		case CRYPTO_CIPHER_MODE_CBC:
+			session->config.Algorithm = CRYP_AES_CBC;
+			ctx->ops.cbc_crypt_hndlr = crypto_stm32_cbc_decrypt;
+			break;
+		case CRYPTO_CIPHER_MODE_CTR:
+			session->config.Algorithm = CRYP_AES_CTR;
+			ctx->ops.ctr_crypt_hndlr = crypto_stm32_ctr_decrypt;
+			break;
+		default:
+			break;
+		}
+	}
+
+	copy_reverse_words((u8_t *)session->key, CRYPTO_STM32_AES_MAX_KEY_LEN,
+			   ctx->key.bit_stream, ctx->keylen);
+
+	session->config.pKey = crypto_stm32_sessions[ctx_idx].key;
+	session->config.DataType = CRYP_DATATYPE_8B;
+	session->config.DataWidthUnit = CRYP_DATAWIDTHUNIT_BYTE;
+
+	ctx->drv_sessn_state = session;
+	ctx->device = dev;
+
+	return 0;
+}
+
+static int crypto_stm32_session_free(struct device *dev,
+				     struct cipher_ctx *ctx)
+{
+	struct crypto_stm32_session *session = CRYPTO_STM32_SESSN(ctx);
+
+	session->in_use = false;
+
+	return 0;
+}
+
+static int crypto_stm32_query_caps(struct device *dev)
+{
+	return CRYP_SUPPORT;
+}
+
+static int crypto_stm32_init(struct device *dev)
+{
+	struct device *clk = device_get_binding(STM32_CLOCK_CONTROL_NAME);
+	struct crypto_stm32_data *data = CRYPTO_STM32_DATA(dev);
+	const struct crypto_stm32_config *cfg = CRYPTO_STM32_CFG(dev);
+
+	__ASSERT_NO_MSG(clk);
+
+	clock_control_on(clk, (clock_control_subsys_t *)&cfg->pclken);
+
+	if (HAL_CRYP_Init(&data->hcryp) != HAL_OK) {
+		LOG_ERR("Initialization error");
+		return -EINVAL;
+	}
+
+	k_sem_init(&data->sem, 1, 1);
+
+	return 0;
+}
+
+static struct crypto_driver_api crypto_enc_funcs = {
+	.begin_session = crypto_stm32_session_setup,
+	.free_session = crypto_stm32_session_free,
+	.crypto_async_callback_set = NULL,
+	.query_hw_caps = crypto_stm32_query_caps,
+};
+
+static struct crypto_stm32_data crypto_stm32_dev_data = {
+	.hcryp = {
+		.Instance = CRYP
```

Comment:
It is a simple remark (relevant only for a code generic on all the families)
CRYP is the common name of the instance and this is fine for F4.
Nevertheless, this wording is different on other families (e-g L5 series). It is called : AES
I do not know why ?!

This is conditionally driven by define in : stm32hxxxx.h :
#define CRYP                ((CRYP_TypeDef *) CRYP_BASE)

Reply:
Yes. This was done on purpose. I decided against `.Instance = (CRYP_TypeDef *) DT_CRYP_0_BASE_ADDRESS`, because I wanted to make clear, that - at this point - only the Cryptographic Accelerator peripheral is supported, the AES hardware encryption accelerator is not. That could easily be added at a later time, I guess. As I don't have suitable hardware at the moment, I decided to concentrate on CRYPT.